File: sdk/python/nucleus_sdk/intent.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Dict, List, Optional

from .models import PodSpec
from .client import ProxyClient


class Intent(str, Enum):
    RESEARCH_WEB = "research_web"
    CODE_REVIEW = "code_review"
    FIX_ISSUE = "fix_issue"
    GENERATE_CODE = "generate_code"
    RELEASE = "release"
    DATABASE_CLIENT = "database_client"
    READ_ONLY = "read_only"
    EDIT_ONLY = "edit_only"
    LOCAL_DEV = "local_dev"
    NETWORK_ONLY = "network_only"
    ORCHESTRATE = "orchestrate"


@dataclass(frozen=True)
class IntentProfile:
    intent: Intent
    profile: str
    description: str
    allowed_ops: List[str]
    gated_ops: List[str]
    notes: Optional[str] = None


INTENT_PROFILES: Dict[Intent, IntentProfile] = {
    Intent.RESEARCH_WEB: IntentProfile(
        intent=Intent.RESEARCH_WEB,
        profile="web_research",
        description="Read files plus web search/fetch.",
        allowed_ops=["read", "glob", "grep", "web_search", "web_fetch"],
        gated_ops=[],
    ),
    Intent.CODE_REVIEW: IntentProfile(
        intent=Intent.CODE_REVIEW,
        profile="code_review",
        description="Read-only repo review with optional web lookup.",
        allowed_ops=["read", "glob", "grep", "web_search"],
        gated_ops=[],
    ),
    Intent.FIX_ISSUE: IntentProfile(
        intent=Intent.FIX_ISSUE,
        profile="fix_issue",
        description="Edit + run tools; approvals required when trifecta is active.",
        allowed_ops=["read", "write", "run", "glob", "grep"],
        gated_ops=["run", "web_fetch", "web_search", "git_push"],
        notes="Exfiltration vectors may require approval when trifecta is complete.",
    ),
    Intent.GENERATE_CODE: IntentProfile(
        intent=Intent.GENERATE_CODE,
        profile="codegen",
        description="Write + run tools without network.",
        allowed_ops=["read", "write", "run", "glob", "grep"],
        gated_ops=[],
    ),
    Intent.RELEASE: IntentProfile(
        intent=Intent.RELEASE,
        profile="release",
        description="Release flow with explicit approvals.",
        allowed_ops=["read", "write", "run", "web_search", "web_fetch"],
        gated_ops=["git_push", "create_pr", "run"],
    ),
    Intent.DATABASE_CLIENT: IntentProfile(
        intent=Intent.DATABASE_CLIENT,
        profile="database_client",
        description="Database CLI only.",
        allowed_ops=["run"],
        gated_ops=[],
    ),
    Intent.READ_ONLY: IntentProfile(
        intent=Intent.READ_ONLY,
        profile="read_only",
        description="Explore files safely; no writes or execution.",
        allowed_ops=["read", "glob", "grep"],
        gated_ops=[],
    ),
    Intent.EDIT_ONLY: IntentProfile(
        intent=Intent.EDIT_ONLY,
        profile="edit_only",
        description="Edit files without execution or network.",
        allowed_ops=["read", "write", "glob", "grep"],
        gated_ops=[],
    ),
    Intent.LOCAL_DEV: IntentProfile(
        intent=Intent.LOCAL_DEV,
        profile="local_dev",
        description="Local development without network.",
        allowed_ops=["read", "write", "run", "glob", "grep"],
        gated_ops=[],
    ),
    Intent.NETWORK_ONLY: IntentProfile(
        intent=Intent.NETWORK_ONLY,
        profile="network_only",
        description="Web-only research with no file access.",
        allowed_ops=["web_search", "web_fetch"],
        gated_ops=[],
    ),
    Intent.ORCHESTRATE: IntentProfile(
        intent=Intent.ORCHESTRATE,
        profile="orchestrator",
        description="Spawn and manage sub-pods. No direct file/command/web access.",
        allowed_ops=["read", "glob", "grep", "create_pod", "list_pods", "pod_status", "pod_logs", "cancel_pod"],
        gated_ops=[],
        notes="Sub-pod permissions bounded by delegation ceiling via monotonic meet.",
    ),
}


class IntentSession:
    def __init__(self, proxy: ProxyClient, profile: IntentProfile, pod_id: Optional[str] = None):
        self._proxy = proxy
        self.profile = profile
        self.pod_id = pod_id

    def describe(self) -> IntentProfile:
        return self.profile

    def read(self, path: str) -> str:
        return self._proxy.read(path)

    def write(self, path: str, contents: str) -> None:
        self._proxy.write(path, contents)

    def run(self, args: List[str], stdin: Optional[str] = None, directory: Optional[str] = None):
        return self._proxy.run(args=args, stdin=stdin, directory=directory)

    def web_fetch(self, url: str, method: Optional[str] = None, headers: Optional[Dict[str, str]] = None, body: Optional[str] = None):
        return self._proxy.web_fetch(url=url, method=method, headers=headers, body=body)

    def web_search(self, query: str, max_results: Optional[int] = None):
        return self._proxy.web_search(query=query, max_results=max_results)

    def glob(self, pattern: str, directory: Optional[str] = None, max_results: Optional[int] = None):
        return self._proxy.glob(pattern=pattern, directory=directory, max_results=max_results)

    def grep(self, pattern: str, path: Optional[str] = None, file_glob: Optional[str] = None, context_lines: Optional[int] = None, max_matches: Optional[int] = None, case_insensitive: Optional[bool] = None):
        return self._proxy.grep(
            pattern=pattern,
            path=path,
            file_glob=file_glob,
            context_lines=context_lines,
            max_matches=max_matches,
            case_insensitive=case_insensitive,
        )


def profile_for_intent(intent: Intent) -> IntentProfile:
    return INTENT_PROFILES[intent]


def pod_spec_for_intent(intent: Intent, work_dir: str = ".", timeout_seconds: int = 3600) -> PodSpec:
    profile = profile_for_intent(intent)
    return PodSpec(work_dir=work_dir, timeout_seconds=timeout_seconds, profile=profile.profile)

File: crates/nucleus-sdk/src/intent.rs
//! Intent system for user-facing permission profiles.
//!
//! [`Intent`] maps high-level goals (e.g., "fix an issue", "review code") to
//! lattice-guard permission profiles. [`IntentSession`] wraps a [`ProxyClient`]
//! with the resolved profile for scoped, type-safe operations.
//!
//! Mirrors the Python SDK's `intent.py`.

use std::collections::HashMap;

use crate::proxy::ProxyClient;
use crate::Error;

/// User-facing intent describing what an agent session will do.
///
/// Each intent maps to a lattice-guard profile name that determines the
/// permission lattice for the session.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Intent {
    /// Web research: read + web_fetch + web_search. No write or exec.
    ResearchWeb,
    /// Code review: read files + glob + grep. No write, no network.
    CodeReview,
    /// Fix an issue: full code editing with trifecta obligations.
    FixIssue,
    /// Generate code: write files in workspace. No network (network-isolated).
    GenerateCode,
    /// Release: git push + PR operations. CI-gated.
    Release,
    /// Database client: network to allowed hosts, no file write.
    DatabaseClient,
    /// Read-only: observe files, no mutations.
    ReadOnly,
    /// Edit-only: write files, no execution or network.
    EditOnly,
    /// Local development: permissive local environment.
    LocalDev,
    /// Network-only: web operations, no filesystem access.
    NetworkOnly,
    /// Orchestrator: manage sub-pods, no direct file/network access.
    Orchestrate,
}

impl Intent {
    /// Get the lattice-guard profile name for this intent.
    pub fn profile_name(&self) -> &'static str {
        match self {
            Intent::ResearchWeb => "web_research",
            Intent::CodeReview => "code_review",
            Intent::FixIssue => "fix_issue",
            Intent::GenerateCode => "codegen",
            Intent::Release => "release",
            Intent::DatabaseClient => "database_client",
            Intent::ReadOnly => "read_only",
            Intent::EditOnly => "edit_only",
            Intent::LocalDev => "local_dev",
            Intent::NetworkOnly => "network_only",
            Intent::Orchestrate => "orchestrator",
        }
    }

    /// Human-readable description of the intent.
    pub fn description(&self) -> &'static str {
        match self {
            Intent::ResearchWeb => "Web research: read + web access, no writes",
            Intent::CodeReview => "Code review: read + search, no writes or network",
            Intent::FixIssue => "Fix issue: full code editing with trifecta obligations",
            Intent::GenerateCode => "Generate code: write files, no network (isolated)",
            Intent::Release => "Release: git push + PR operations, CI-gated",
            Intent::DatabaseClient => "Database client: network to allowed hosts, no file write",
            Intent::ReadOnly => "Read-only: observe files, no mutations",
            Intent::EditOnly => "Edit-only: write files, no execution or network",
            Intent::LocalDev => "Local development: permissive local environment",
            Intent::NetworkOnly => "Network-only: web operations, no filesystem access",
            Intent::Orchestrate => "Orchestrator: manage sub-pods only",
        }
    }

    /// All available intents.
    pub fn all() -> &'static [Intent] {
        &[
            Intent::ResearchWeb,
            Intent::CodeReview,
            Intent::FixIssue,
            Intent::GenerateCode,
            Intent::Release,
            Intent::DatabaseClient,
            Intent::ReadOnly,
            Intent::EditOnly,
            Intent::LocalDev,
            Intent::NetworkOnly,
            Intent::Orchestrate,
        ]
    }
}

/// Resolved profile for an intent, including the permission lattice.
#[derive(Debug, Clone)]
pub struct IntentProfile {
    /// The intent this profile was derived from.
    pub intent: Intent,
    /// Lattice-guard profile name.
    pub profile_name: &'static str,
    /// Human-readable description.
    pub description: &'static str,
    /// Resolved permission lattice.
    pub lattice: lattice_guard::PermissionLattice,
}

impl IntentProfile {
    /// Resolve an intent to its full profile including lattice.
    pub fn resolve(intent: Intent) -> Result<Self, Error> {
        let profile_name = intent.profile_name();
        let policy = nucleus_spec::PolicySpec::Profile {
            name: profile_name.to_string(),
        };
        let lattice = policy.resolve().map_err(|e| {
            Error::Config(format!(
                "failed to resolve profile '{}': {}",
                profile_name, e
            ))
        })?;

        Ok(Self {
            intent,
            profile_name,
            description: intent.description(),
            lattice,
        })
    }

    /// Check which operations are allowed at autonomous level.
    pub fn allowed_operations(&self) -> Vec<lattice_guard::Operation> {
        use lattice_guard::{CapabilityLevel, Operation};

        let ops = [
            (Operation::ReadFiles, self.lattice.capabilities.read_files),
            (Operation::WriteFiles, self.lattice.capabilities.write_files),
            (Operation::RunBash, self.lattice.capabilities.run_bash),
            (Operation::WebFetch, self.lattice.capabilities.web_fetch),
            (Operation::WebSearch, self.lattice.capabilities.web_search),
            (Operation::GitPush, self.lattice.capabilities.git_push),
            (Operation::GitCommit, self.lattice.capabilities.git_commit),
            (Operation::CreatePr, self.lattice.capabilities.create_pr),
            (Operation::GlobSearch, self.lattice.capabilities.glob_search),
            (Operation::GrepSearch, self.lattice.capabilities.grep_search),
        ];

        ops.iter()
            .filter(|(_, level)| {
                matches!(level, CapabilityLevel::Always | CapabilityLevel::LowRisk)
            })
            .map(|(op, _)| *op)
            .collect()
    }

    /// Check which operations have approval obligations (gated by trifecta enforcement).
    pub fn gated_operations(&self) -> Vec<lattice_guard::Operation> {
        self.lattice.obligations.approvals.iter().copied().collect()
    }
}

/// A session bound to an intent profile, wrapping a [`ProxyClient`].
///
/// Provides the same operations as `ProxyClient` but scoped to the intent's
/// permission profile. The tool-proxy enforces the actual lattice — this is
/// a client-side convenience for intent-aware usage.
///
/// # Example
///
/// ```rust,no_run
/// use nucleus_sdk::{Nucleus, Intent, HmacAuth};
///
/// # async fn example() -> nucleus_sdk::Result<()> {
/// let nucleus = Nucleus::builder()
///     .proxy_url("http://127.0.0.1:8080")
///     .auth(HmacAuth::new(b"secret", None))
///     .build()?;
///
/// let session = nucleus.intent(Intent::CodeReview).await?;
/// let source = session.read("src/lib.rs").await?;
/// // session.write(...) would be denied by the tool-proxy (CodeReview is read-only)
/// # Ok(())
/// # }
/// ```
pub struct IntentSession {
    proxy: ProxyClient,
    /// The resolved intent profile for this session.
    pub profile: IntentProfile,
    /// Pod ID if this session was created via nucleus-node.
    pub pod_id: Option<String>,
}

impl IntentSession {
    /// Create a new intent session wrapping a proxy client.
    pub fn new(proxy: ProxyClient, profile: IntentProfile, pod_id: Option<String>) -> Self {
        Self {
            proxy,
            profile,
            pod_id,
        }
    }

    /// Get the underlying proxy client.
    pub fn proxy(&self) -> &ProxyClient {
        &self.proxy
    }

    // -- Delegated operations --

    /// Read a file's contents.
    pub async fn read(&self, path: &str) -> Result<String, Error> {
        self.proxy.read(path).await
    }

    /// Write contents to a file.
    pub async fn write(&self, path: &str, contents: &str) -> Result<(), Error> {
        self.proxy.write(path, contents).await
    }

    /// Run a command.
    pub async fn run(
        &self,
        args: &[&str],
        stdin: Option<&str>,
        directory: Option<&str>,
    ) -> Result<crate::proxy::RunOutput, Error> {
        self.proxy.run(args, stdin, directory).await
    }

    /// Search for files matching a glob pattern.
    pub async fn glob(
        &self,
        pattern: &str,
        directory: Option<&str>,
        max_results: Option<u32>,
    ) -> Result<crate::proxy::GlobOutput, Error> {
        self.proxy.glob(pattern, directory, max_results).await
    }

    /// Search file contents with a regex pattern.
    pub async fn grep(
        &self,
        pattern: &str,
        path: Option<&str>,
        file_glob: Option<&str>,
        context_lines: Option<u32>,
        max_matches: Option<u32>,
        case_insensitive: Option<bool>,
    ) -> Result<crate::proxy::GrepOutput, Error> {
        self.proxy
            .grep(
                pattern,
                path,
                file_glob,
                context_lines,
                max_matches,
                case_insensitive,
            )
            .await
    }

    /// Fetch a URL.
    pub async fn web_fetch(
        &self,
        url: &str,
        method: Option<&str>,
        headers: Option<&HashMap<String, String>>,
        body: Option<&str>,
    ) -> Result<serde_json::Value, Error> {
        self.proxy.web_fetch(url, method, headers, body).await
    }

    /// Grant pre-approval for an operation.
    pub async fn approve(
        &self,
        operation: &str,
        count: u32,
        expires_at_unix: Option<u64>,
        nonce: Option<&str>,
    ) -> Result<serde_json::Value, Error> {
        self.proxy
            .approve(operation, count, expires_at_unix, nonce)
            .await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_all_intents_resolve() {
        for intent in Intent::all() {
            let profile = IntentProfile::resolve(*intent);
            assert!(
                profile.is_ok(),
                "Intent {:?} should resolve, got: {:?}",
                intent,
                profile.err()
            );
        }
    }

    #[test]
    fn test_intent_profile_names_match() {
        let cases = [
            (Intent::ResearchWeb, "web_research"),
            (Intent::CodeReview, "code_review"),
            (Intent::FixIssue, "fix_issue"),
            (Intent::GenerateCode, "codegen"),
            (Intent::Release, "release"),
            (Intent::DatabaseClient, "database_client"),
            (Intent::ReadOnly, "read_only"),
            (Intent::EditOnly, "edit_only"),
            (Intent::LocalDev, "local_dev"),
            (Intent::NetworkOnly, "network_only"),
            (Intent::Orchestrate, "orchestrator"),
        ];

        for (intent, expected) in cases {
            assert_eq!(intent.profile_name(), expected);
        }
    }

    #[test]
    fn test_code_review_is_read_only() {
        let profile = IntentProfile::resolve(Intent::CodeReview).unwrap();
        let allowed = profile.allowed_operations();

        // Code review should allow read, glob, grep
        assert!(allowed.contains(&lattice_guard::Operation::ReadFiles));
        assert!(allowed.contains(&lattice_guard::Operation::GlobSearch));
        assert!(allowed.contains(&lattice_guard::Operation::GrepSearch));

        // Should NOT allow write, run_bash, git push
        assert!(!allowed.contains(&lattice_guard::Operation::WriteFiles));
        assert!(!allowed.contains(&lattice_guard::Operation::RunBash));
        assert!(!allowed.contains(&lattice_guard::Operation::GitPush));
    }

    #[test]
    fn test_fix_issue_has_trifecta_awareness() {
        let profile = IntentProfile::resolve(Intent::FixIssue).unwrap();

        // Fix issue has broad capabilities — check that gated ops exist
        // (trifecta enforcement adds approval obligations)
        let _gated = profile.gated_operations();
        let allowed = profile.allowed_operations();

        // Should allow read and write
        assert!(allowed.contains(&lattice_guard::Operation::ReadFiles));
        assert!(allowed.contains(&lattice_guard::Operation::WriteFiles));
    }

    #[test]
    fn test_codegen_no_network() {
        let profile = IntentProfile::resolve(Intent::GenerateCode).unwrap();
        let allowed = profile.allowed_operations();

        // Codegen should allow write
        assert!(allowed.contains(&lattice_guard::Operation::WriteFiles));

        // Should NOT allow web fetch (network-isolated)
        assert!(!allowed.contains(&lattice_guard::Operation::WebFetch));
    }

    #[test]
    fn test_orchestrator_pod_management() {
        let profile = IntentProfile::resolve(Intent::Orchestrate).unwrap();

        // Orchestrator has limited direct ops but manages sub-pods
        // The key assertion is that it resolves without error
        assert_eq!(profile.profile_name, "orchestrator");
    }

    #[test]
    fn test_all_intents_count() {
        assert_eq!(Intent::all().len(), 11);
    }
}
